### ISSUE TITLE
Allow to stop execution of a resolver from a plugin

### ISSUE THREAD

**Issue by dotansimha** [graphql-hive/envelop#383]

**Comment by santino** [comment]
Thanks for raising the placeholder.

To summarise the minimum requirement is to arbitrarily force a resolver to return a given value and terminate, or better avoid, execution of the original resolver function.
This would be something similar to `setResultAndStopExecution` available as part of `onExecute`.

Another way to achieve the same result is to replace the original resolver function with a custom function that can just return an arbitrary value and since it replaces the original it prevents that from being executed.
This would be something similar to `setExecuteFn` available as part of `onExecute`.

**Comment by dotansimha** [comment]
@santino few thoughts:
1. We can do `skipResolverExecution` and skip the resolver code itself, but keep the rest of the plugins running.
2. We can do `stopResolver` that stops and skip the rest of the "before" plugins calls.
3. We can do `stopResolver` that stops and skip the rest of the "before" and "after" plugin calls.

Which one do you think we should start with? Because we have 3 different flows here: Before plugins, the resolver itself and then After plugins.

**Comment by santino** [comment]
Thanks for your thoughts, I was also having a look at the code which helps with the thinking.
I think we still want to keep the plugins tracking the resolver, so we want to skip the resolver code itself only; but not prevent the resolver from being hit.

In the code you have
`const originalFn: GraphQLFieldResolver<any, any> = field.resolve || defaultFieldResolver;`

Maybe we can pass a function that takes precedence over the original `field.resolve` in there.

Alternatively, a flag like `preventResolverFunction` can also work, and when you have this flag you can set result value to null maybe, turning this `let result = await originalFn(root, args, context, info);` into this `let result = preventResolverFunction ? null : await originalFn(root, args, context, info);`.

Then later on you can still continue to replace result like its already happening with `setResult: newResult => { result = newResult; },`

Does it makes sense?

**Comment by dotansimha** [comment]
> Thanks for your thoughts, I was also having a look at the code which helps with the thinking.
> I think we still want to keep the plugins tracking the resolver, so we want to skip the resolver code itself only; but not prevent the resolver from being hit.
> 
> In the code you have
> `const originalFn: GraphQLFieldResolver<any, any> = field.resolve || defaultFieldResolver;`
> 
> Maybe we can pass a function that takes precedence over the original `field.resolve` in there.
> 
> Alternatively, a flag like `preventResolverFunction` can also work, and when you have this flag you can set result value to null maybe, turning this `let result = await originalFn(root, args, context, info);` into this `let result = preventResolverFunction ? null : await originalFn(root, args, context, info);`.
> 
> Then later on you can still continue to replace result like its already happening with `setResult: newResult => { result = newResult; },`
> 
> Does it makes sense?

I guess it depends if you want to do in your plugin:

```
replaceResolverFunction(myFunc)
```

and then everything will continue the same way, just with the modified resolver.

Or, you want to do:

```
stopAndReplaceResolverResult(resolverFakeReturnValue)
```

Feels like option #1 is more robust because it will allow you to do #2 as well, and it can play nicely with the rest of the plugins you are adding.

**Comment by santino** [comment]
yes, option 1 is enough for my personal usecase.
I will be ok with that only, but maybe you want to add both anyway, and again this would be similar to `setExecuteFn` and `setResultAndStopExecution` as you have for Execute

**Comment by dotansimha** [comment]
> yes, option 1 is enough for my personal usecase.

Cool, done in: https://github.com/dotansimha/envelop/pull/400

> I will be ok with that only, but maybe you want to add both anyway, and again this would be similar to `setExecuteFn` and `setResultAndStopExecution` as you have for Execute

Keeping that on-hold, I want to think about short-circuit in general (see https://github.com/dotansimha/envelop/issues/359)

**Comment by dotansimha** [comment]
Available in `@envelop/core@0.4.0`

**Comment by santino** [comment]
Great stuff, thanks Dotan!!!
